### PR TITLE
Remove loop check from differentiation (trac4454)

### PR DESCRIPTION
**Pull request by hkiel**
ther is still a maxIteration check.
reduces running time tremendously

### CODE REVIEW

**Comment by lochel** [comment]
I am not sure if we should remove the check completely. However, I agree that the current solution is way too slow according to your analysis. As far as I know, @sjoelund has introduced it to improve the messages that are generated by the max-iteration-check.

I guess it could be implemented more efficiently using hash sets instead of lists, because it is only checked if an element is stored or not.

@sjoelund Any comment?

**Comment by perost** [comment]
You could check the expression stack for duplicates only if the max iteration limit is reached. Then you'd still get the same error message, but the only overhead in case of no error would be building the list of expressions.

**Comment by lochel** [comment]
@perost Sounds like a good idea. On one hand there are cases where this is slower, compared to the hash set approach but on the other hand it should be faster as long as there is no loop.

**Comment by perost** [comment]
@lochel I assumed this was a hard error that would cause the compilation to abort, but otherwise it might be slower letting it run into the iteration limit as you say. But if the compilation is going to abort though it doesn't really matter if that case is slightly slower, having low overhead in the normal case is more important in my opinion.

**Comment by lochel** [comment]
@perost I agree. I just wanted to state it here for the sake of completeness.

**Comment by hkiel** [comment]
A loop in differentiation seems to be a hard error:
```Error.addInternalError("Differentiation failed...");```
So, shall we go for 
a) building the list and checking at iterationMax
or
b) using HashSet?

**Comment by lochel** [comment]
It would be interesting to know how big the overhead is for the hash set approach. If that is too big then I think we should do the check only at iterationMax.

File: Compiler/BackEnd/Differentiate.mo [hunk 1]
```diff
@@ -537,11 +534,6 @@ algorithm
     fail();
   end if;
 */
-  if listMember(inExp, inExpStack) then
```
Comment (hkiel) [inline Compiler/BackEnd/Differentiate.mo]: @lochel @perost I see that maximum iteration check is currently disabled

**Comment by hkiel** [comment]
Just for the statistics: not creating the list saves 125MB in ScalableTestSuite.Electrical.BreakerCircuits.ScaledExperiments.BreakerNetwork_N_640 i.e. 15*10^6 elements in the lists...

**Comment by sjoelund** [comment]
Do you think calling GC.free would help?

⁣Sent from BlueMail ​

**Comment by hkiel** [comment]
Not for the list.

**Comment by hkiel** [comment]
Can somebody construct a test case where a loop in differentiateExp() happens???

**Comment by hkiel** [comment]
Can we make a test run on hudson for this PR? Would be interesting to see if it breaks something and if the performance increase can be seen for small models as well.

**Comment by OpenModelica-Hudson** [comment]
Wrong hash OMCompiler e24e32354c5d3613d50833ccc1fbc0b8b6e0b05d e24e323

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2017-06-28_15-55-07](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/4359/console).

**Comment by hkiel** [comment]
I implemented loop check with HashSetExp.
Using HashSetExp instead of list<DAE.Exp> makes no difference both in speed and memory.

So, do we really need the loop checking or can we skip this completely?
@sjoelund commented the maxIteration check out with the following comment:
> This check does not seem to be necessary since looking through the stack of expression seems to stop iteration in most cases, and you get a spam of messages from this check.

Another option would be to add a flag for loop checking (checkDifferentiationLoop). But what would be the default?

**Comment by lochel** [comment]
It would be good to see some examples before drawing a conclusion. @sjoelund, can you point to some examples that trigger this check?

**Comment by sjoelund** [comment]
A HashSet is a very expensive datatype creating a large array by default. You should only use it when you set the size of the set to an appropriate size beforehand. Otherwise, use an AvlSet.

But yes, we could simply have a maximum number of iterations instead as long as it is implemented properly (high enough limit that you don't get messages, and only 1 message per failed differentiation).

**Comment by sjoelund** [comment]
@lochel - this mostly triggered errors when the module was implemented improperly; it's more of a sanity check to see if the compiler is doing something wrong.

**Comment by hkiel** [comment]
@sjoelund if it's only checking for errors in the compiler, then I would skip the loop-check completely, like it is done in this PR.

The maximum iteration number depends on the model size. Currently it is defined as a fixed number (20), but should be set to some pre-calculated number. For ScalableTestSuite.Electrical.BreakerCircuits.ScaledExperiments.BreakerNetwork_N_640 I set it to 700 (640 was too small). I would go for 2048, but then as the function is recursive, I do not know how many iterations the stack can hold...

**Comment by hkiel** [comment]
As @sjoelund stated this check is a sanity check to see if compiler is doing something wrong. So I'm going to merge this PR now.
How can I trigger a run of ScalableTestSuite?

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2017-07-07_10-00-18](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/4382/console).

**Pull request closed without merging**